### PR TITLE
Feat/opcode sar

### PR DESCRIPTION
**Pull request by lispc**

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/sar.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+impl<F: Field> ExecutionGadget<F> for SarGadget<F> {
+    const NAME: &'static str = "SAR";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::SAR;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        let a = cb.query_word();
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/sar.rs:33]: add simple comments, what is a and what is b

File: zkevm-circuits/src/evm_circuit/execution/sar.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    #[test]
+    fn sar_gadget_simple() {
+        test_ok(OpcodeId::SAR, 0x02FF.into(), 0x1.into());
+        test_ok(
+            OpcodeId::SAR,
+            Word::from_big_endian(&[255u8; 32]),
+            0x73.into(),
+        );
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/sar.rs:105]: add a testcase with "shift" bigger than 256

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 3]
```diff
@@ -1570,3 +1570,290 @@ impl<F: Field> AbsWordGadget<F> {
...
+            .zip(a64s_lo.iter())
+            .map(|(lt, val)| lt.assign(region, offset, F::from_u128(*val), F::from_u128(p_lo)))
+            .collect::<Result<Vec<_>, _>>()?;
+        Ok(())
+    }
+}
+
+#[inline]
+fn is_neg(x: &U256) -> bool {
+    127 < x.to_le_bytes()[31]
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:1858]: x < 0 or x.is_negative()?

File: zkevm-circuits/src/evm_circuit/execution/sar.rs (hunk 2 above)
Comment (LuozhuZhang) [reply zkevm-circuits/src/evm_circuit/execution/sar.rs]: refactor test mod

**Comment by lispc** [comment]
LGTM.

todo: 

1. i leave one comment
2. fix ci

**Comment by lispc** [comment]
moved to https://github.com/scroll-tech/zkevm-circuits/pull/190

**Pull request closed without merging**